### ISSUE TITLE
[core] Request for help on the _rs2rs() method

### ISSUE THREAD

**Issue by mnewnham** [ADOdb/ADOdb#777]
### TLDR
Why does the method `_rs2rs()` completely replace the calling object instead of setting up the recordset in the calling object. If you have any thoughts or ideas at all, please comment,  

### Overview
I'm looking for opinions on a a specific piece of code in the core `adodb.inc.php` file. in particular, I am looking at `ADOConnection::_execute()`. 

There is a section of code relating to building a recordset if the `ADORecordset::_numOfRows` variable is zero. Generally this only occurs in database drivers that do not support a native "number of rows" returned  method. Whilst few, they do however exist. In particular, this exists in the IBM DB2 driver  when using forward cursors (default behavior) 

 So in ADOConnection::_execute() we see the code:
````php

if ($rs->_numOfRows <= 0) 
{
    global $ADODB_COUNTRECS;
   
    if ($ADODB_COUNTRECS) {
        if (!$rs->EOF) {
            $rs = $this->_rs2rs($rs,-1,-1,!is_array($sql)); <---- This is the line
            $rs->_queryID = $this->_queryID;
        } else
            $rs->_numOfRows = 0;
    }
}

````
So if the driver cannot return the number of rows, and `$ADODB_COUNTRECS` is set to true, then the method '_rs2rs()' is called. This method is defined in ADOConnection:

````php 
/**
* Convert a database recordset to an array recordset.
*
* Input recordset's cursor should be at beginning, and old $rs will be closed.
*
* @param ADORecordSet $rs     the recordset to copy
* @param int          $nrows  number of rows to retrieve (optional)
* @param int          $offset offset by number of rows (optional)
* @param bool         $close
*
* @return ADORecordSet_array|ADORecordSet|bool the new recordset
*/
function &_rs2rs(&$rs,$nrows=-1,$offset=-1,$close=true) {
    if (! $rs) {
        $ret = false;
        return $ret;
    }
    etc....... 
}
````
So the first parameter is the recordset, and it is called by reference, so any changes we have made to `$rs` in the function would be returned to the main recordset. But you will also notice that the entire recordset object is overwritten by the return value from `_rs2rs()` as well. So what does this mean, for example using the IBM DB2 driver:
 
1. We create an ADOConnection using the driver DB2. This creates an `ADOConnection_db2` object. 
2. We execute an SQL statement, and the result set is initialized in a `ADORecordset_db2` object. Although there may be say, 10 records in the result, the value of `ADORecordset::_numOfRows`  is zero. 
3. We therefore enter the code in `ADOConnection::_execute()` where  `_rs2rs()` is called, and because of the design of the code, as described above,  the original recordset object, in this case `ADORecordset_db2` is  replaced by the parent `ADORecordset` object. 
 
This means that any driver specific methods in the recordset are no longer available to the class, or are replaced by the basic parent methods. This effectively kills any driver specific functionality. 

 The question is why therefore does this need to replace the class, instead of just updating the recordset. Incidentally, there appears to be no problem with simply updating the recordset, i.e. replacing

````php
$rs = $this->_rs2rs($rs,-1,-1,!is_array($sql));
````
With
````php
$this->_rs2rs($rs,-1,-1,!is_array($sql));
````

**Comment by dregad** [comment]
I actually started looking at this one last night, but was too tired and went to bed before replying...

Did you find the answer to your question ? I would think it should be fine to remove the assignment, but did not run any tests to confirm that.
